### PR TITLE
Add rocketmq configuration items, Add orderly message. 

### PR DESCRIPTION
**Pull request by x-shadow-man**
Add the RocketMQ MetaData configuration to be consistent with the configuration of the RocketMQ GO Client.
Support for RocketMQ's orderly message capabilities.
Re-implement the Subscribe startup logic.

### CODE REVIEW

File: pubsub/rocketmq/metadata.go [hunk 1]
```diff
@@ -85,25 +175,44 @@ func (s *rocketMQMetaData) Decode(in interface{}) error {
...
-	if rMetaData.GroupName != "" {
-		logger.Warn("pubsub.rocketmq: metadata property 'groupName' has been deprecated - use 'producerGroup' instead. See: https://docs.dapr.io/reference/components-reference/supported-pubsub/setup-rocketmq/")
-	}
-
-	if rMetaData.SendTimeOut != 0 {
-		logger.Warn("pubsub.rocketmq: metadata property 'sendTimeOut' has been deprecated - use 'sendTimeOutSec' instead. See: https://docs.dapr.io/reference/components-reference/supported-pubsub/setup-rocketmq/")
-		if rMetaData.SendTimeOutSec == 0 {
-			rMetaData.SendTimeOutSec = rMetaData.SendTimeOut / 1000000
+		if rMetaData.InstanceName == "" {
+			rMetaData.InstanceName = metadata.Properties["instance"]
```
Comment (berndverst) [inline pubsub/rocketmq/metadata.go:192]: I think this should be `instanceName` ?

Also please define all these property keys as constants.

File: pubsub/rocketmq/metadata.go [hunk 2]
```diff
@@ -85,25 +175,44 @@ func (s *rocketMQMetaData) Decode(in interface{}) error {
...
+		}
+		if rMetaData.ConsumerGroupName == "" {
+			rMetaData.ConsumerGroupName = metadata.Properties["consumerGroup"]
```
Comment (berndverst) [inline pubsub/rocketmq/metadata.go:195]: Make this string key a constant.

File: pubsub/rocketmq/metadata.go [hunk 3]
```diff
@@ -14,68 +14,158 @@ limitations under the License.
...
+	HashQueueSelector       QueueSelectorType = "hash"
+	RandomQueueSelector     QueueSelectorType = "random"
+	ManualQueueSelector     QueueSelectorType = "manual"
+	RoundRobinQueueSelector QueueSelectorType = "roundRobin"
+	DaprQueueSelector       QueueSelectorType = "dapr"
+)
+
+// RocketMQ Go Client Options
 type rocketMQMetaData struct {
-	// rocketmq Credentials
```
Comment (berndverst) [inline pubsub/rocketmq/metadata.go:41]: Please do not remove these comments here - they are very helpful for maintaining the code and metadata. As a result it's also important you don't change the order of the metadata here.

**Review by berndverst — CHANGES_REQUESTED** [review]

File: pubsub/rocketmq/metadata.go (hunk 3 above)
Comment (jinjianfeng-chn) [reply pubsub/rocketmq/metadata.go:41]: ok, iI'll fill in the comments

File: pubsub/rocketmq/metadata.go (hunk 2 above)
Comment (jinjianfeng-chn) [reply pubsub/rocketmq/metadata.go:195]: ok

File: pubsub/rocketmq/metadata.go (hunk 1 above)
Comment (jinjianfeng-chn) [reply pubsub/rocketmq/metadata.go:192]: InstanceName has the same effect as instance. Instance is for better compatibility

Comment (jinjianfeng-chn) [reply pubsub/rocketmq/metadata.go:192]: ok, I will fill in the comments

File: pubsub/rocketmq/metadata.go [hunk 4]
```diff
@@ -14,68 +14,166 @@ limitations under the License.
...
-	GroupName     string `mapstructure:"groupName"`
-	ProducerGroup string `mapstructure:"producerGroup"`
-	NameSpace     string `mapstructure:"nameSpace"`
-	// consumer group rocketmq's subscribers
-	ConsumerGroup     string `mapstructure:"consumerGroup"`
-	ConsumerBatchSize int    `mapstructure:"consumerBatchSize"`
+	// rocketmq instance name, it will be registered to the broker
+	InstanceName string `mapstructure:"instanceName"`
+	// consumer group name
+	ConsumerGroupName string `mapstructure:"consumerGroupName"`
```
Comment (ItalyPaleAle) [inline pubsub/rocketmq/metadata.go:51]: This seems to have changed from before. Before, it was `groupName`. If you want to rename keys, you must make sure that the old one is still supported as an alias (for example: make `groupName` an alias of `consumerGroupName`, and show a warning in the logs if people use `groupName`)

File: pubsub/rocketmq/metadata.go [hunk 5]
```diff
@@ -14,68 +14,166 @@ limitations under the License.
...
+	// producer group name
+	ProducerGroupName string `mapstructure:"producerGroupName"`
```
Comment (ItalyPaleAle) [inline pubsub/rocketmq/metadata.go:53]: Same, this seems to havebeen renamed from `producerGroup`. Please maintain the old one as alias

File: pubsub/rocketmq/metadata.go [hunk 6]
```diff
@@ -14,68 +14,166 @@ limitations under the License.
...
-	AccessKey  string `mapstructure:"accessKey"`
-	SecretKey  string `mapstructure:"secretKey"`
-	NameServer string `mapstructure:"nameServer"`
-	// Deprecated: use ProducerGroup instead.
-	GroupName     string `mapstructure:"groupName"`
-	ProducerGroup string `mapstructure:"producerGroup"`
-	NameSpace     string `mapstructure:"nameSpace"`
-	// consumer group rocketmq's subscribers
-	ConsumerGroup     string `mapstructure:"consumerGroup"`
-	ConsumerBatchSize int    `mapstructure:"consumerBatchSize"`
```
Comment (ItalyPaleAle) [inline pubsub/rocketmq/metadata.go:51]: This option seems to have been removed? Please, we should maintain backwards-compatiblity

File: pubsub/rocketmq/metadata.go [hunk 7]
```diff
@@ -14,68 +14,166 @@ limitations under the License.
...
+	// rocketmq Credentials
+	AccessKey     string `mapstructure:"accessKey"`
+	SecretKey     string `mapstructure:"secretKey"`
+	SecurityToken string `mapstructure:"securityToken"`
+	// retry times to send msg to broker
 	Retries int `mapstructure:"retries"`
-	// Deprecated: send msg timeout to connect rocketmq's broker, nanoseconds
-	SendTimeOut int `mapstructure:"sendTimeOut"`
-	// send msg timeout to connect rocketmq's broker, seconds
-	SendTimeOutSec int `mapstructure:"sendTimeOutSec"`
```
Comment (ItalyPaleAle) [inline pubsub/rocketmq/metadata.go:61]: This has been removed too?

Comment (ItalyPaleAle) [reply pubsub/rocketmq/metadata.go:61]: Per below, "sendTimeOut" should have been removed (deprecated), while "sendTimeOutSec" needs to stay.

File: pubsub/rocketmq/metadata.go [hunk 8]
```diff
@@ -85,25 +183,55 @@ func (s *rocketMQMetaData) Decode(in interface{}) error {
...
-func parseRocketMQMetaData(metadata pubsub.Metadata, logger logger.Logger) (*rocketMQMetaData, error) {
-	rMetaData := getDefaultRocketMQMetaData()
+const (
+	keyInstance         string = "instance"
+	KeyConsumerGroup    string = "consumerGroup"
+	KeyProducerGroup    string = "producerGroup"
+	KeyConsumeFromWhere string = "consumeFromWhere"
+	KeyConsumeOrder     string = "consumeOrder"
+	KeyQueueSelector    string = "queueSelector"
+	KeySendMsgTimeout   string = "sendMsgTimeout"
```
Comment (ItalyPaleAle) [inline pubsub/rocketmq/metadata.go:193]: Please maintain compatibility with "sendTimeOutSec"

File: pubsub/rocketmq/rocketmq.go [hunk 9]
```diff
@@ -16,120 +16,247 @@ package rocketmq
...
+	if r.metadata.ConsumerModel != "" {
+		switch strings.ToLower(r.metadata.ConsumerModel) {
+		case "broadcasting":
+			opts = append(opts, mqc.WithConsumerModel(mqc.BroadCasting))
+		case "clustering":
+			opts = append(opts, mqc.WithConsumerModel(mqc.Clustering))
+		default:
+			r.metadata.ConsumerModel = "Clustering"
+			opts = append(opts, mqc.WithConsumerModel(mqc.Clustering))
+			r.logger.Warnf("%s Consumer Model[%s] is error, expected [BroadCasting, Clustering], "+
```
Comment (ItalyPaleAle) [inline pubsub/rocketmq/rocketmq.go:152]: `"%s Consumer Model[%s] is invalid: expected [broadcasting, clustering]; we will use default model [clustering]"`

File: pubsub/rocketmq/rocketmq.go [hunk 10]
```diff
@@ -16,120 +16,247 @@ package rocketmq
...
+		case "consumefromfirstoffset":
+			opts = append(opts, mqc.WithConsumeFromWhere(mqc.ConsumeFromFirstOffset))
+		case "consumefromtimestamp":
+			opts = append(opts, mqc.WithConsumeFromWhere(mqc.ConsumeFromTimestamp))
+		default:
+			r.metadata.FromWhere = "ConsumeFromLastOffset"
+			opts = append(opts, mqc.WithConsumeFromWhere(mqc.ConsumeFromLastOffset))
+			r.logger.Warnf("%s Consumer FromWhere[%s] is error, "+
+				"expected [ConsumeFromLastOffset, ConsumeFromFirstOffset, ConsumeFromTimestamp], "+
+				"we will use default value [ConsumeFromLastOffset]", r.name, r.metadata.FromWhere)
```
Comment (ItalyPaleAle) [inline pubsub/rocketmq/rocketmq.go:169]: `"%s Consumer FromWhere[%s] is error, expected [ConsumeFromLastOffset, ConsumeFromFirstOffset, ConsumeFromTimestamp], we will use default value [ConsumeFromLastOffset]"

File: pubsub/rocketmq/rocketmq.go [hunk 11]
```diff
@@ -16,120 +16,247 @@ package rocketmq
...
+		}
+	}
+	if r.metadata.ConsumeMessageBatchMaxSize > 0 {
+		opts = append(opts, mqc.WithConsumeMessageBatchMaxSize(r.metadata.ConsumeMessageBatchMaxSize))
+	}
+	if r.metadata.MaxReconsumeTimes > 0 {
+		opts = append(opts, mqc.WithMaxReconsumeTimes(r.metadata.MaxReconsumeTimes))
+	}
+	if r.metadata.AutoCommit != "" {
+		if utils.IsTruthy(r.metadata.AutoCommit) {
```
Comment (ItalyPaleAle) [inline pubsub/rocketmq/rocketmq.go:190]: From:

```
		if utils.IsTruthy(r.metadata.AutoCommit) {
			opts = append(opts, mqc.WithAutoCommit(true))
		} else {
			opts = append(opts, mqc.WithAutoCommit(false))
		}
```

To:

```
		opts = append(opts, mqc.WithAutoCommit(utils.IsTruthy(r.metadata.AutoCommit)))
```

File: pubsub/rocketmq/rocketmq.go [hunk 12]
```diff
@@ -16,120 +16,247 @@ package rocketmq
...
-		opts = append(opts, mqp.WithGroupName(r.metadata.GroupName))
+	if r.metadata.ProducerGroupName != "" {
+		opts = append(opts, mqp.WithGroupName(r.metadata.ProducerGroupName))
 	}
-	if r.metadata.ProducerGroup != "" {
-		opts = append(opts, mqp.WithGroupName(r.metadata.ProducerGroup))
-	}
-	if r.metadata.NameServerDomain != "" {
-		opts = append(opts, mqp.WithNameServerDomain(r.metadata.NameServerDomain))
+	if r.metadata.NameServer != "" {
```
Comment (ItalyPaleAle) [inline pubsub/rocketmq/rocketmq.go:217]: This is also in `setUpConsumer`: can you please create a helper function for this?

File: pubsub/rocketmq/rocketmq.go [hunk 13]
```diff
@@ -146,208 +273,198 @@ func (r *rocketMQ) Features() []pubsub.Feature {
...
 
-func (r *rocketMQ) Publish(req *pubsub.PublishRequest) error {
-	r.logger.Debugf("rocketmq publish topic:%s with data:%v", req.Topic, req.Data)
-	msg := newRocketMQMessage(req)
-
-	publishBo := backoff.NewExponentialBackOff()
-	publishBo.InitialInterval = 100 * time.Millisecond
-	bo := backoff.WithMaxRetries(publishBo, 3)
-	bo = backoff.WithContext(bo, r.ctx)
-	return retry.NotifyRecover(
```
Comment (ItalyPaleAle) [inline pubsub/rocketmq/rocketmq.go:157]: Why did you remove the backoff?

File: pubsub/rocketmq/rocketmq.go [hunk 14]
```diff
@@ -16,120 +16,247 @@ package rocketmq
...
-	producerLock sync.RWMutex
-	consumer     mq.PushConsumer
-	consumerLock sync.RWMutex
-
-	ctx    context.Context
-	cancel context.CancelFunc
+	name          string
+	metadata      *rocketMQMetaData
+	producer      mq.Producer
+	producerLock  sync.RWMutex
```
Comment (ItalyPaleAle) [inline pubsub/rocketmq/rocketmq.go:71]: Looks like you're not using RLocks, so can be changed to `sync.Mutex`

File: pubsub/rocketmq/rocketmq.go [hunk 15]
```diff
@@ -16,120 +16,247 @@ package rocketmq
...
+	consumer      mq.PushConsumer
+	consumerLock  sync.RWMutex
```
Comment (ItalyPaleAle) [inline pubsub/rocketmq/rocketmq.go:73]: Looks like you're not using RLocks, so can be changed to `sync.Mutex`

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: pubsub/rocketmq/metadata.go (hunk 4 above)
Comment (jinjianfeng-chn) [reply pubsub/rocketmq/metadata.go:51]: ok, I'll keep the groupName as the alias for consumerGroupName.

File: pubsub/rocketmq/metadata.go (hunk 7 above)
Comment (jinjianfeng-chn) [reply pubsub/rocketmq/metadata.go:61]: ok，I'll drop the sendTimeOut renaming.

File: pubsub/rocketmq/metadata.go (hunk 8 above)
Comment (jinjianfeng-chn) [reply pubsub/rocketmq/metadata.go:193]: ok

File: pubsub/rocketmq/metadata.go (hunk 6 above)
Comment (jinjianfeng-chn) [reply pubsub/rocketmq/metadata.go:51]: ConsumerBatchSize is renamed to PullBatchSize.
 I'm compatible with ConsumerBatchSize.

File: pubsub/rocketmq/rocketmq.go (hunk 9 above)
Comment (jinjianfeng-chn) [reply pubsub/rocketmq/rocketmq.go:152]: ok, thanks

File: pubsub/rocketmq/rocketmq.go (hunk 10 above)
Comment (jinjianfeng-chn) [reply pubsub/rocketmq/rocketmq.go:169]: ok

File: pubsub/rocketmq/rocketmq.go (hunk 11 above)
Comment (jinjianfeng-chn) [reply pubsub/rocketmq/rocketmq.go:190]: ok

File: pubsub/rocketmq/rocketmq.go (hunk 12 above)
Comment (jinjianfeng-chn) [reply pubsub/rocketmq/rocketmq.go:217]: ok, i will create a helper function for parse NameServer

File: pubsub/rocketmq/rocketmq.go (hunk 13 above)
Comment (jinjianfeng-chn) [reply pubsub/rocketmq/rocketmq.go:157]: The RocketMQ Client will retry if it fails. The retry times are the values of metadata.Retries.
We don't need to retry here.

File: pubsub/rocketmq/rocketmq.go (hunk 14 above)
Comment (jinjianfeng-chn) [reply pubsub/rocketmq/rocketmq.go:71]: ok

**Comment by berndverst** [comment]
I noticed we are using go-rocketmq 2.1.1-rc2 but there is version 2.1.1 which is very different.

Could you also help getting us into a stable version instead of the RC?

**Comment by jinjianfeng-chn** [comment]
> 

v2.1.1-rc2 is the latest compatible version. The stable version of v2.1.1 is not compatible, so I rolled back to v2.1.0
